### DISCUSSION TITLE
After Update Product/:id/productAttributeValues Endpoint returns nothing

### DISCUSSION THREAD

**Discussion by marcus-at-localhost** [atrocore/atrocore#588] [Q&A] [answered]
Hi,

![image](https://user-images.githubusercontent.com/1337223/210775512-031188bf-3b3a-4621-84b5-aef66e793ca7.png)

i just updated the PIM (after a week or so) and the Endpoint returns an empty 200 

The error log says:

```
[2023-01-05T11:35:03.719747+00:00] Log.ERROR: Uncaught Exception Error: "Cannot use object of type Pim\Entities\ProductAttributeValue as array" at xxx/vendor/atrocore/pim/app/Services/Product.php line 716 {"exception":"[object] (Error(code: 0): Cannot use object of type Pim\\Entities\\ProductAttributeValue as array at xxx/vendor/atrocore/pim/app/Services/Product.php:716)"} []
```

I have no idea how to fix it. If I check the same endpoint in the PIM App in the browser, it works as expected.

![image](https://user-images.githubusercontent.com/1337223/210776117-eab4b30a-757c-47e8-a245-9718d18b866f.png)

But with Postman it crashes.

![image](https://user-images.githubusercontent.com/1337223/210775898-5b9a2a85-3ec6-464f-a975-c63349fd046b.png)

I assume it has something with the language to do:
https://github.com/atrocore/atropim/blob/master/app/Services/Product.php#L716

With every API call I send `language: en_US` as header, that seems to work for other Endpoints.

Please advise. Thank you.

## Edit:

I debugged it and it comes down to this change:
```php
        $headerLanguage = self::getHeader('language');
		$GLOBALS['log']->debug('Language', (array)$headerLanguage);
		
        // filtering via header language
        if (!empty($headerLanguage)) {
            foreach ($records as $id => $pav) {
				$GLOBALS['log']->debug('Records', (array)$pav->language);
                if ($pav->language !== $headerLanguage) {
                    //unset($records[$id]);
                }
            }
        }
```
The error message was clear, it's `$pav->language` not an array `$pav['language']` that gave an expected response, but incomplete, obviously most records are being filtered out.

But the Debug Statements say this:

```
[2023-01-05T12:15:28.623053+00:00] Log.DEBUG: Language ["en_US"] []
[2023-01-05T12:15:28.623817+00:00] Log.DEBUG: Records ["main"] []
```

The headerLanguage is `en_US` and the record language is `main`.

Do you have any insight into this behavior? Is this something wrong in my setup?

**Comment by rratsun** [comment, marked as answer]
Hello,

We will fix this in PIM 1.6.1. The stable release should be on this or next week.

**Reply by marcus-at-localhost** [reply]
Thanks! Are there any potential side effects to my hack/fix?

**Reply by rratsun** [reply]
No :)

But the stable release should be today.

**Reply by marcus-at-localhost** [reply]
I ran the update and it solved the initial issue with the object access. Thank you.

I still have to comment out the `unset($records[$id])` line because the `header.language:en_US` is compared to `$record.language:main` and it returns only one attribute (which has `$record.language:en_US`). How gets `$record.language:main` resolved so it returns whatever language I request via header.language?

**Reply by rratsun** [reply]
We added one more improvement. Please, update the system. After update described problem should be solved.

**Reply by marcus-at-localhost** [reply]
I just updated from 1.6.1. to 1.6.7 and unfortunately the attributes are missing, when I call `/Product/:entityId/productAttributeValues`. The only one that's retuned is one with an actual language setting .

![image](https://user-images.githubusercontent.com/1337223/212035619-6b57b23b-7276-4be1-8685-06baa3ed16b0.png)

In the App API calls all attributes are being returned and the language key has `main` in it.

![image](https://user-images.githubusercontent.com/1337223/212036088-1062eec1-91f4-4c5e-829a-7b15223bb9a0.png)

Has anything changed in how to access the API related to language? That must have happened in the last days?

I don't know how to debug this further :/.

**Reply by marcus-at-localhost** [reply]
Ok, if I don't send the Language in the Header of the API call, then I get everything back. But when and why did that change? Was that documented?
How can I now get the correct language version of the attribute labels?


 I try to stay on top of the updates and read the commits and changelogs, but it's often not very descriptive and especially API changes should be communicated more clearly.

**Reply by marcus-at-localhost** [reply]
I narrowed it down to changes in the API (see https://github.com/atrocore/atrocore/discussions/597) that have nothing to do with the original issue here. So this thread is answered.
